Only watch tls secrets

raffis opened the pull request:
## Description
Emissary has quite some performance problems on our cluster. It's using 1 full core  and over 7G memory without any requests. Basically on standby.
![Screenshot from 2022-09-22 09-48-43](https://user-images.githubusercontent.com/2376735/191775597-f4090a3a-83bd-4357-a243-27020e14a485.png)

Profiling:
![Screenshot from 2022-09-22 09-39-11](https://user-images.githubusercontent.com/2376735/191775633-dc26209d-d333-4657-a6f8-5119669f7d4d.png)

The problem is that ambassador basically ends up reconciling non stop because it watches quite a big resource list including secrets, ingresses, ep and so on.
On a big cluster this can be quite cpu/mem heavy.
In our case we have ~10.000 secrets on the cluster and emissary globally (not single namespace) will read all  10.000 secrets.
On top of that some secrets change quite often which triggers basically an infinite reconcile loop.

The solution for this is:
* Remove ingress access for emissary from the clusterrole (we only use ambassador mappings)
* This pr, as ambassador should only read tls secrets. I don't see a reason why ambassador should read other secrets. (This reduces the number of secrets from 10.000 to like 30 in our case).

After those changes emissary uses like 120m/400Mi on average.


If non tls secrets are used for any reason there should be a possibility to declare a resource filter instead.


**Note**: I also updated the dev guide as the things there do not work as stated.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `CHANGELOG.md`.

   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations

 - [x] This is unlikely to impact how Ambassador performs at scale.

   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability

 - [ ] My change is adequately tested.

   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points

 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

 - [ ] The changes in this PR have been reviewed for security concerns and adherence to security best practices.

haq204 commented on cmd/entrypoint/interesting_types.go:
```diff
@@ -80,9 +80,9 @@ func GetInterestingTypes(ctx context.Context, serverTypeList []kates.APIResource
-		"K8sSecrets": {{typename: "secrets.v1."}},                              // New in Kubernetes 0.16.0 (2015-04-28) (v1beta{1..3} before that)
+		"Services":   {{typename: "services.v1."}},                                         // New in Kubernetes 0.16.0 (2015-04-28) (v1beta{1..3} before that)
+		"Endpoints":  {{typename: "endpoints.v1.", fieldselector: endpointFs}},             // New in Kubernetes 0.16.0 (2015-04-28) (v1beta{1..3} before that)
+		"K8sSecrets": {{typename: "secrets.v1.", fieldselector: "type=kubernetes.io/tls"}}, // New in Kubernetes 0.16.0 (2015-04-28) (v1beta{1..3} before that)
```
Not all folks put tls secrets as `kubernetes.io/tls`. you could create a normal secret and use the `tls.crt` and `tls.key` fields. So this could end up being a breaking change.

raffis replied:
fair enough, how about the the 2nd proposal then to add the ability to configure a filter? It doesnt make sense to load and watch 10.000 resources in the case only a handful are required.
Where would that config go? Module crd?

raffis replied:
or actually from an env only as the module res would need to be loaded first otherwise (if thats not the case, didnt check the code yet)

haq204 reviewed (changes requested):
My main concern is that this could end up breaking some customers if they don't store tls secrets as `kubernetes.io/tls`

Also we merged https://github.com/emissary-ingress/emissary/pull/4488 which also helps with constant re-configurations and will be available in v3.2 and v2.4

AliceProxy commented:
> Thanks for the contribution!
> 
> My main concern is that this could end up breaking some users if they don't store tls secrets as `kubernetes.io/tls`
> 
> Also we merged #4488 which also helps with constant re-configurations and will be available in v3.2 and v2.4

Also note that support for certificate revocation lists loads the CRL configuration for the feature from from non-tls secrets. I believe we also do watch for opaque secrets for validation contexts.

raffis commented:
> Also we merged https://github.com/emissary-ingress/emissary/pull/4488 which also helps with constant re-configurations and will be available in v3.2 and v2.4


v3.2 does not fix it. It also consumes like 10G of memory just because it reads the entire list of secrets from the cluster.

raffis commented:
> As for the idea to configure a filter, I think that works. In terms of where to place it, Module is a pretty old resource so we probably don't want to add new things in there. So I would probably prefer as an env var, to configure which secrets to watch based on label selectors? But that shouldn't be any different than `AMBASSADOR_LABEL_SELECTOR` https://www.getambassador.io/docs/emissary/latest/topics/running/environment/#ambassador_label_selector. I'll admit that we should make this option more visible in the performance and scaling section of the docs but have you tried setting this option and labeling your secrets to see if that resolves your issue?


I was actually not aware that env exists, this might solve it but it sounds unpractical since this involves labeling everything from mappings, services, secrets and so on right? 


What would be nice is to optionally specify the resourcegroup and support multiple selectors. Something like:
```
AMBASSADOR_LABEL_SELECTOR=secrets.v1:mylabel=myvalue,services.v1:mylabel=test
```

In my case I could then label my two secrets and define this env like:
```
AMBASSADOR_LABEL_SELECTOR=secrets.v1:emissary-secret=yes
```

Otherwise I would need to change like 50 repos and label services and so on as well.
This change might even be backwards compatible, didn't think that through yet though.

haq204 commented:
in order to specify label selectors per resource, let's create a new env var idk call it `AMBASSADOR_WATCHER_LABEL_SELECTOR` and we can mark the `AMBASSADOR_LABEL_SELECTOR` as deprecated to be remove in a future version as the formats will be different.

We could have the same env var support the two different formats but I figure that would make the parsing logic more complicated so it might be more clear to have them as two separate env vars.

How does that sound?

raffis commented:
New functionality implemented, docs here: https://github.com/datawire/ambassador-docs/pull/700
I also added the same for the field selector (which was not even documented before).

Two comments, 
* When I run tests with `make gotest` it takes ages and fail, seeing some timeouts like `testutil_fake_test_test.go:86: Get timed out!` in the output and a lot of other output. This looks unrelated to my change. The tests added should work. They failed first time I implemented it and this was visible before all the follow up errors.
* I see that the env AMBASSADOR_LABEL_SELECTOR is also used in python/*, honestly no clue what the python stuff is for, and ARCHITECTURE.md is empty where I hoped to get informed about this. Not sure if something needs to be changed there.

However I tested the new feature on our cluster and it works so far.

haq204 commented:
> New functionality implemented, docs here: [datawire/ambassador-docs#700](https://github.com/datawire/ambassador-docs/pull/700) I also added the same for the field selector (which was not even documented before).
> 
> Two comments,
> 
> * When I run tests with `make gotest` it takes ages and fails eventually, seeing some timeouts like `testutil_fake_test_test.go:86: Get timed out!` in the output and a lots of other output. This looks unrelated to my change. The tests added should work. They failed first time I implemented it and this was visible before all the follow up errors.
> * I see that the env AMBASSADOR_LABEL_SELECTOR is also used in python/*, honestly no clue what the python stuff is for, and ARCHITECTURE.md is empty where I hoped to get informed about this. Not sure if something needs to be changed there.
> 
> However I tested the new feature on our cluster and it works so far.

I pulled down your changes to check. `make gotest` also runs integration tests (and also runs in CI which unfortunately doesn't run correctly for community contributions due to insufficient permissions but we're actively working to fix that).

In order to properly run them you will need to set the following env vars:
```
DEV_REGISTRY # Docker registry to push/pull local builds to
DEV_KUBECONFIG #path to a valid kubeconfig file. Integrations tests will spawn a k3s cluster container
```

At this point, I can bring over your PR so that we can run CI on your changes.

for the python stuff, that is old legacy code. I don't think we need to update that but lets see what CI says.

haq204 reviewed (commented):
once the comment is addressed I can bring over your changes so that we can run CI on it.

raffis commented:
> I pulled down your changes to check. `make gotest` also runs integration tests (and also runs in CI which unfortunately doesn't run correctly for community contributions due to insufficient permissions but we're actively working to fix that).
> 
> In order to properly run them you will need to set the following env vars:
> 
> ```
> DEV_REGISTRY # Docker registry to push/pull local builds to
> DEV_KUBECONFIG #path to a valid kubeconfig file. Integrations tests will spawn a k3s cluster container
> ```
> 
> For the python stuff, that is old legacy code. I don't think we need to update that.

Thanks for the explanation 👍🏻

raffis commented on cmd/entrypoint/interesting_types.go:
```diff
@@ -52,6 +62,42 @@ func GetQueries(ctx context.Context, interestingTypes map[string]thingToWatch) [
 	return queries
 }
 
+func getWatchFieldSelector(resourcegroup string) string {
```
Mhm I reckon thats micro optimization as this function is only called at bootstrapping as far as I can tell. But yeah lets change it 👍🏻

raffis commented:
I changed some bits and it supports now selectors without the apiversion for the resourcegroup.

raffis commented:
@haq204 can we run the ci job?

raffis commented:
Sorry to bother you but can I please 🙏🏻 have an update on this? @haq204 @AliceProxy

haq204 commented:
Hi @raffis sorry for the delay got busy with other things. 

So change wise it looks good to me but I brought this up to the rest of the Maintainers team and we were discussing it internally. It turns out that this might actually be a regressions from all the way back to 1.14 Legacy Mode. As far as my understanding goes it's not supposed to watch for all secrets and only for this it cares about. 

Given this information, this seems to suggest that the proper approach now is to, well fix the regression rather than adding another env var toggle.

@LukeShu @kflynn I think your thoughts would be constructive to this.

raffis commented:
Whats the status with this? We still have to run my fork because of this bug.
This fix is generic and not only about secrets, you can filter all kind of resources this way.

iomarmochtar commented:
I face the same issue as what explained by @raffis and seems this changes can solve it, waiting forward for any progress in this MR @haq204

raffis commented:
This problem still exists, just tested again with the latest released
